worker_tasks/categorize.py

```diff
@@ -46,109 +81,374 @@
 
-        # Ignore single characters
-        if len(category) < 2:
-            return False
+    def _get_relatedness_score(self, article_1_links, article_2_links):
+        """Calculate overlap of two Wikipedia articles: 2 * [# of shared links] / [total # of links]"""
 
-        # Ignore stop words
-        if category.lower() in STOP_WORDS:
+        shared_links_count = 0
+        total_links_count = len(article_1_links) + len(article_2_links)
+
+        for article_link in article_1_links:
+            weight = 1
+            if "Template" in article_link or \
+                "Wikipedia" in article_link or \
+                "Help" in article_link:
+                weight = 0.1
+            if "Category" in article_link:
+                weight = 1.5
+
+            if article_link in article_2_links:
+                shared_links_count += weight
+
+        return 2.0 * shared_links_count / total_links_count
+
+
+    def _get_best_keywords(self, communities, original_keywords):
+        """Return all keywords from clusters that contain a satisfactory number of the original keywords"""
+
+        best_keywords = set()
+        keyword_counts = Counter(original_keywords)
+
+        for cluster in communities.subgraphs():
+            vertex_names = map(lambda x: x['name'], cluster.vs)
+            
+            matching_vertices_count = sum(map(lambda x: keyword_counts[x] if x in keyword_counts else 0, vertex_names))
+
+            cluster_score = float(matching_vertices_count) / len(cluster.vs)
+
+            if cluster_score > 1:
+                best_keywords.update(vertex_names)
+
+        if len(best_keywords) < 2:
+            best_original_keywords = keyword_counts.most_common(2)
+            best_keywords.update([keyword[0] for keyword in best_original_keywords])
+        
+        return [keyword.title() for keyword in best_keywords] if best_keywords else [""]
+
+
+    def _is_viable_candidate(self, phrase):
+        
+        # Ignore 1-character phrases
+        if len(phrase) < 2:
             return False
 
         # Ignore numbers
         try:
-            float(category)
+            phrase = phrase.replace(',', '')
+            float(phrase)
             return False
         except ValueError:
             pass
 
-        return True
\ No newline at end of file
+        words = phrase.split()
+
+        # Ignore stop words
+        if words[0].lower() in STOP_WORDS:
+            return False
+
+        # Ignore non-title trigrams
+        if len(words) == 3:
+            if words[2].lower() in STOP_WORDS:
+                return False
+
+            titled_words = map(lambda x: x.istitle(), words)
+            if titled_words not in [[True, False, True], [True, True,True]]:
+                return False 
+
+        return True
+
+
+    def _scrape_wiki_links(self, titles):
+        unscraped_titles = filter(lambda x: x not in self._memoized_article_links, titles)
+
+        for i in xrange(0, len(unscraped_titles), 5):
```

Comment:
Wiki didn't have any recommendations; I just came up with 5. I think 10 was too much and I didn't want to push it for fear that we crash the scraper.